rebac-admin-backend/v1/groups_validation_types.go

```diff
@@ -0,0 +1,70 @@
+// Copyright 2024 Canonical Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+package v1
+
+import (
+	"github.com/canonical/identity-platform-admin-ui/rebac-admin-backend/v1/resources"
+)
+
+func validateGroup(value *resources.Group) error {
+	if len(value.Name) == 0 {
+		return NewRequestBodyValidationError("empty group name")
+	}
+	return nil
+}
+
+func validateGroupEntitlementsPatchRequestBody(value *resources.GroupEntitlementsPatchRequestBody) error {
+	if len(value.Patches) == 0 {
+		return NewRequestBodyValidationError("empty patch array")
+	}
+	return validateSlice[resources.GroupEntitlementsPatchItem](value.Patches, validateGroupEntitlementsPatchItem)
+}
+
+func validateGroupEntitlementsPatchItem(value *resources.GroupEntitlementsPatchItem) error {
+	if err := validateEntityEntitlement(&value.Entitlement); err != nil {
+		return err
+	}
+	return validateStringEnum("op", value.Op, resources.GroupEntitlementsPatchItemOpAdd, resources.GroupEntitlementsPatchItemOpRemove)
+}
+
+func validateEntityEntitlement(value *resources.EntityEntitlement) error {
+	if len(value.EntitlementType) == 0 {
+		return NewRequestBodyValidationError("empty entitlement type")
+	}
+	if len(value.EntityName) == 0 {
+		return NewRequestBodyValidationError("empty entity name")
+	}
+	if len(value.EntityType) == 0 {
+		return NewRequestBodyValidationError("empty entity type")
+	}
+	return nil
+}
+
+func validateGroupIdentitiesPatchRequestBody(value *resources.GroupIdentitiesPatchRequestBody) error {
+	if len(value.Patches) == 0 {
+		return NewRequestBodyValidationError("empty patch array")
+	}
+	return validateSlice[resources.GroupIdentitiesPatchItem](value.Patches, validateGroupIdentitiesPatchItem)
+}
```

Comment:
this could be implemented on types generation, see https://github.com/deepmap/oapi-codegen/issues/227

basically using the [extensions](https://github.com/deepmap/oapi-codegen?tab=readme-ov-file#extensions) in the specific the `x-oapi-codegen-extra-tags`, this would add an extra yaml file, but you would have validation tags/annotations in the struct itself

Reply:
Yeah, I checked the extensions, but what I don't like about them is that they clutter the spec with generator-specific tags.

Reply:
that is fair, but at the same time right now we have types without effective embedded validation which doesn't sound ideal either

Reply:
That's fair, too. But the solution would be to change the code generator instead of touching the spec.